src/OrderedMsgChain.js

```diff
@@ -0,0 +1,101 @@
+import debugFactory from 'debug'
+import Heap from 'heap'
+import { Errors, MessageLayer } from 'streamr-client-protocol'
+
+import GapFillFailedError from './errors/GapFillFailedError'
+
+const debug = debugFactory('StreamrClient::OrderedMsgChain')
+const { MessageRef } = MessageLayer
+
+const DEFAULT_GAPFILL_TIMEOUT = 5000
+const MAX_GAP_REQUESTS = 10
+
+export default class OrderedMsgChain {
+    constructor(publisherId, msgChainId, inOrderHandler, gapHandler, gapFillTimeout = DEFAULT_GAPFILL_TIMEOUT) {
+        this.publisherId = publisherId
+        this.msgChainId = msgChainId
+        this.inOrderHandler = inOrderHandler
+        this.gapHandler = gapHandler
+        this.lastReceivedMsgRef = null
+        this.gapFillTimeout = gapFillTimeout
+        this.queue = new Heap((msg1, msg2) => {
+            return msg1.getMessageRef().compareTo(msg2.getMessageRef())
+        })
+    }
+
+    add(unorderedStreamMessage) {
+        const msgRef = unorderedStreamMessage.getMessageRef()
+        if (this.lastReceivedMsgRef && msgRef.compareTo(this.lastReceivedMsgRef) <= 0) {
+            // Prevent double-processing of messages for any reason
+            debug('Already received message: %o, lastReceivedMsgRef: %d. Ignoring message.', msgRef, this.lastReceivedMsgRef)
+            return
+        }
+
+        if (this._isNextMessage(unorderedStreamMessage)) {
+            this._process(unorderedStreamMessage)
+            this._checkQueue()
+        } else {
+            if (!this.gap) {
+                this._scheduleGap()
+            }
+            this.queue.push(unorderedStreamMessage)
+        }
+    }
+
+    addError(err) {
```

Comment:
I think this class doesn't necessarily need to know about the structure of the error. You could dig out the streamMessage from the error outside this class and then refactor this to be something like
```js
markMessageExplicitly(streamMessage) {
    if (this._isNextMessage(streamMessage)) {
        this.lastReceivedMsgRef = err.getMessageRef()
    }
}
```